### DISCUSSION TITLE
curious power consumption

### DISCUSSION THREAD

**Discussion by roha-github** [meshtastic/firmware#4173] [Ideas]
The power consumption varies greatly depending on the release and setting. 

I took a few measurements and got some unexpected results. In my opinion, the behavior is transferable to other devices and ESP32 models, even if the Heltec V2 is no longer officially supported. I measured with a USB consumption meter.

| Version | Idle mA | Down mA | Settings |
|---|---|---|---|
| 2.2.23.5672e68 | 44 | 6 | display-, position- |
| 2.2.23.5672e68 | 43 | 6 | display-, position-, serial- |
| 2.2.23.5672e68 | 17 | 14 | display-, position-, serial-, powersave+ |
| 2.2.23.5672e68 | 32 | 37 | display-, position-, serial-, powersave+, bluetooth- |
| 2.2.23.5672e68 | 60 | 24 | display-, position-, bluetooth- |
| 2.2.23.5672e68 | 60 | 22 | display-, position-, bluetooth-, serial- |
| 2.3.4.ea61808 | 56 | 37 | display-, position- |

Activating Power Saving reduces the power consumption during reception due to Light Sleep, but increases the power consumption during shutdown. It gets even stranger if you deactivate Bluetooth, in which case the power consumption actually increases instead of decreasing. As of release 2.3.x, more power is generally consumed. 

Of course, this is not necessarily a runtime error, but it shows that the states of the consumptions are not (cleanly) defined in the same way for different options. It is also difficult to automatically test the consumption depending on different settings.

--- 
Edit: I noticed something else - after flashing firmware 2.2.23, the node consumes approx. 60mA when idle. If I disconnect the device from the USB port and reconnect it, the node consumes approx. 44mA when idle.

This could indicate that the ESP32 is calibrated differently during booting and has an influence on the power consumption.

https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/RF_calibration.html

I would like to understand what influences the different power consumption, because ESP32 wants to be used as a sensor with disposable batteries.

---
Edit 2: I couldn't get my head around it, so I downloaded the source code for 2.2.23 from Github and compiled it unchanged with VSCode (1.90.2) and PlatformIO (Core 6.1.15 / Home 3.4.4) - result 12mA after shutdown vs. 6mA with the precompiled binary. Apparently the compiler also has an influence on the consumption.

**Comment by GUVWAF** [comment]
> Apparently the compiler also has an influence on the consumption.

I highly doubt this. This seems to suggest your measurements are not consistent. 
What does "idle" mean in this case and is this an average during a certain time? Was it receiving (or even transmitting) and processing packets during this time, connecting and talking to a client, etc.?

**Reply by roha-github** [reply]
Idle means that the node is waiting to receive data but is not receiving any. The consumption is certainly an average value from the multimeter. How do you determine the consumption?

The display does not jump around, but is displayed permanently. Have you tried to understand this?

**Reply by GUVWAF** [reply]
How do you know it's not receiving any? And how do you make sure it's also not transmitting anything?

"The consumption is certainly an average value from the multimeter." and "The display does not jump around, but is displayed permanently." is what confuses me. So if it's a constant value, the "average value" is then just what it displays? I don't expect to see a constant value.

Anyway, with statements like this: "I noticed something else - after flashing firmware 2.2.23, the node consumes approx. 60mA when idle. If I disconnect the device from the USB port and reconnect it, the node consumes approx. 44mA when idle." it's clear that comparing different firmware versions like this is basically useless, as the measurement error exceeds any difference you report.

**Reply by roha-github** [reply]
First of all, thank you for addressing my question about the difference in power consumption.

Your last answer shows me that you have apparently not understood my experimental setup. Perhaps you could try it first without doubting my competence.

During the development of source code, I measure with a USB multimeter. It shows the current consumption in mA and mW as well as mAh and mWh over time.

"How do you know it's not receiving any? And how do you make sure it's also not transmitting anything?"

During the measurements, I use serial debug and monitor the status LED. I am the only user of Meshtastic in my immediate vicinity. When I make such measurements, no other nodes are on. I measure about two minutes after switching on the node.

Before we discuss the content any further, I would like you to use the above-mentioned device and USB multimeter, then reflash the firmware and then disconnect it from the USB and measure it again. As long as you are only speculating, we cannot clarify the content of the question.

Perhaps someone else has made similar measurements with other dev boards.

With TLoRa I have reduced the current consumption from 1.9mA to 150uA in deep sleep. I would like to understand why turning off functionalities generates more power consumption. 

Have you investigated the consumption of different releases and features/configurations? What were the results and how should it be measured differently?

**Reply by GUVWAF** [reply]
No, I've not done this because it's hard to do this reliably. I also don't have a Heltec V2. Previously people have measured power consumption over many hours of runtime.

Some work is being done to do more fine-grained measurements based on what the node is doing at each time: https://github.com/meshtastic/firmware/issues/4136 

No, I didn't understand your experimental setup because you didn't explain it fully in your original post. It's still not clear to me  how the values in your table are derived. Is it from the mAh reported over a certain time and if so, how long?

**Reply by roha-github** [reply]
Thanks for the tip about PowerMon, I'm looking into it right now.

As long as I don't have a long-term measurement here, it doesn't seem to make sense for us to continue discussing this.

Maybe someone else will observe similar supposed power consumption depending on the features. Then we can get back into it and possibly clarify whether something could be optimized in terms of power consumption.
